### PR TITLE
feat(extraction): support tree artifacts in bazel extractor

### PR DESCRIPTION
**Pull request by nbeloglazov**
Usually artifacts in bazel are regular files. But sometimes it can be directories
(tree artifacts): https://docs.bazel.build/versions/master/glossary.html#artifact
In this case extractor has to recursively include all files in such directories.

Had to refactor unit test as now extractor performs file IO so need to initialize
all input files by creating empty files.

### CODE REVIEW

**Comment by shahms** [comment]
It looks like the go-module build fails as well as gofmt checks.

**Comment by nbeloglazov** [comment]
What is go-module build? I don't think I have access to the build logs so no idea why it failed or how to run it locally. With gofmt should be a bit easier.

**Comment by shahms** [comment]
Ah, we want the Go code to be buildable outside of bazel so also run: `go test -mod=readonly ./kythe/...` during CI. In this case, it's complaining about being unable to find the "treeset" module.

**Comment by nbeloglazov** [comment]
I updated this PR. Should be good now. @shahms can you trigger presubmits again?

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
+// Package treeset provides functions for extracting targets that use bazel treesets.
```
Comment (shahms) [inline kythe/go/extractors/bazel/treeset/treeset.go:1]: This file needs a "Kythe Authors" copyright header.

Comment (nbeloglazov) [reply kythe/go/extractors/bazel/treeset/treeset.go:17]: Of course it does... Done.

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+// ExpandDirectories returns the list of files contained in the provided paths.
+//
+// Paths can be individual files or directories. If it's a directory - all files
+// within that directory added to the result.
+func ExpandDirectories(paths []string) []string {
+	nps := make([]string, 0)
+	for _, root := range paths {
+		files, err := ListSources(root)
+		if err != nil {
+			fmt.Printf("couldn't list files for %s: %s\n", root, err)
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:97]: log.Printf("WARNING: couldn't...

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	return missingInputs.Elements()
+}
+
+// ExpandDirectories returns the list of files contained in the provided paths.
+//
+// Paths can be individual files or directories. If it's a directory - all files
+// within that directory added to the result.
+func ExpandDirectories(paths []string) []string {
+	nps := make([]string, 0)
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:93]: var nps []string

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+			dir := filepath.Dir(file)
+			for dir != "" {
+				if inputsSet.Contains(dir) {
+					missingInputs.Add(file)
+					break
+				}
+				dir = filepath.Dir(dir)
+			}
+			if dir == "" {
+				fmt.Printf("Couldn't find an input for %s\n", file)
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:80]: log.Printf("WARNING: couldn't...

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+// Some bazel rules use tree artifacts for the inputs to the compiler. These are directories that
+// Bazel expands to files when the action is run. Consequently, the list of inputs that Bazel has
+// only contains the tree artifact directory. This function reports the files that are (1) required,
+// (2) not included in bazel's inputs, and (3) have their parent directory included in Bazel's
+// inputs.
+func FindMissingTreeInputs(inputs []string, requiredFiles stringset.Set) []string {
+	missingInputs := stringset.New()
+	inputsSet := stringset.New(inputs...)
+	for file := range requiredFiles {
+		if !inputsSet.Contains(file) {
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:70]: Switch this to

```
if inputsSet.Contains(file) {
  continue
}
dir := filepath.Dir(file)
....
```

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	"path/filepath"
+
+	"bitbucket.org/creachadair/stringset"
+)
+
+// ListSources returns the source files underneath path. If path is a file, it returns a set with
+// a single element. If path is a directory, it returns a set containing all the files
+// (recursively) under path.
+func ListSources(arg string) (stringset.Set, error) {
+	f, err := os.Open(arg)
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:32]: Use kythe/go/platform/vfs.Open (you'll need to plumb Contexts around the package).

**Review by schroederc — CHANGES_REQUESTED** [review]

File: kythe/go/extractors/bazel/treeset/treeset.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+// Some bazel rules use tree artifacts for the inputs to the compiler. These are directories that
+// Bazel expands to files when the action is run. Consequently, the list of inputs that Bazel has
+// only contains the tree artifact directory. This function reports the files that are (1) required,
+// (2) not included in bazel's inputs, and (3) have their parent directory included in Bazel's
+// inputs.
+func FindMissingTreeInputs(inputs []string, requiredFiles stringset.Set) []string {
+	missingInputs := stringset.New()
+	inputsSet := stringset.New(inputs...)
+	for file := range requiredFiles {
+
```
Comment (schroederc) [inline kythe/go/extractors/bazel/treeset/treeset.go:69]: nit: remove blank line

**Review by schroederc — APPROVED** [review]

**Pull request merged**